Support streaming operations

roidrage opened the pull request:
Support running mapred and keys in a streaming fashion, i.e. run the callback with every chunk of data received.

This is a bit of a work in progress, and I still have to add tests, currently they're broken for me locally, but I wanted to bring it up for discussion first.

roidrage committed:
Allow tasks to be streaming, for mapred and keys.

roidrage committed:
Fix typo.

nlf commented:
I like this idea, but maybe an event emitter would be better? This is definitely a quicker and easier approach, but I feel like firing an event is better suited for this task than repeatedly running the callback.

roidrage commented:
I'm all for it. riak-js would use an event emitter for this.

Should the client of riakpbc provide one or should riakpbc return one?

nlf commented:
I'm thinking the client returns one, something like..

``` js
var keys = client.getKeys({ bucket: 'test' });
keys.on('data', function (chunk) {
  console.log(chunk); // ['key1', 'key2']
});
keys.on('end', function (chunk) {
  console.log('done!');
});
```

Call the callback if it's there, and return an event emitter regardless. That's actually a pattern that all of the calls _could_ follow, but I'm not sure if I would like how that works for the non-streaming messages. Let's start with just keys and mapreduce for now, and maybe I'll play around with that idea and see how it looks for the other calls.

roidrage commented:
Sounds reasonable to me. riak-js for example only uses an event emitter for these two operations :)

I'll look into it!

roidrage committed:
Use EventEmitters for streaming.

roidrage commented:
First attempt to use EventEmitters for this approach. Going to test this better.

roidrage committed:
Merge branch 'master' into mm-streaming-mapreduce

Conflicts:
	index.js

roidrage committed:
Rework event triggers.

Triggers the decoded response to the emitter instead of the growing
reply object.

roidrage commented:
I reworked this a bit to actually send the single chunks.

There's one thing I've been wondering about: by returning an EventEmitter, is there a remote chance that an event could be missed if no listener is registered yet? Parts of riak-js have adopted the approach that the client needs to call a `start()` function on the emitter object to explicitly start the operation after the listeners have been registered.

nlf commented:
Yes, there is a small chance events could be missed. That's something that I believe the Stream interface in node 0.10 aimed to resolve, but I haven't really played with it much. I'm reasonably sure it shouldn't be an issue as is, but it might be worth further testing.

nlf commented on index.js:
```diff
@@ -152,8 +164,14 @@ RiakPBC.prototype.setBucket = function (params, callback) {
-RiakPBC.prototype.getKeys = function (params, callback) {
-    this.makeRequest('RpbListKeysReq', params, callback, true);
+RiakPBC.prototype.getKeys = function (params, streaming, callback) {
+    if (streaming) {
```
I'd prefer if this defaulted to false.. i.e.

``` js
if (typeof streaming === 'function') {
    callback = streaming;
    streaming = false;
}
```

roidrage commented:
Would it be reasonable then to use the same behaviour and have the client call `start()` explicitly when the events are attached? Personally not too sure, I'd be fine with leaving it like this for now until it becomes an issue.

I'll look into the suggested change.

nlf commented:
I think leave it as is for now, it's likely an edge case that the listener won't be created quickly enough to catch all of the messages.

roidrage commented:
Sounds good, thanks!

roidrage committed:
Default streaming to false.

Ensure backward compatibility for the interface.

roidrage committed:
Merge branch 'master' into mm-streaming-mapreduce

pgte commented on index.js:
```diff
@@ -1,6 +1,7 @@
 var net = require('net'),
     protobuf = require('protobuf.js'),
     butils = require('butils'),
+    EventEmitter = require('events').EventEmitter,
```
You shouldn't use an event emitter, instead should use [a readable stream](http://nodejs.org/api/stream.html#stream_class_stream_readable):

```
Readable = require('stream').Readable;
```

roidrage replied:
Does that exist in Node.js 0.8?

pgte replied:
No, if Node 0.8 compatibility is important to you, you would have to do something like:

``` js
Stream = require('stream');
```

and then implement a readable:

``` js
var stream = new Stream();
stream.readable = true;
```

and emit `data` events on it (as you were doing).

I'd use the better streams2 API if you can use Node 0.10.
If not, there is the `readable_stream` shim: https://github.com/isaacs/readable-stream
, which you can use like this:
https://github.com/isaacs/http-duplex-client/blob/master/client.js#L7

roidrage replied:
As this is all for riak-js to use, I'd prefer to support Node.js 0.8 for the time being.

I'll look into readable streams in general, but will ship a first version with EventEmitters for now.

Thanks!

pgte replied:
OK, just a heads up: it won't be pipable if you use a plain event emitter.

Later, once this lands, and since I'm interested in this feature, I can take a look at this streaming feature and make it compatible with both 0.8 and 0.10.

roidrage replied:
Sounds good, and the help and pointers are much appreciated!

The riak-js interface for this particular part has always been using an EventEmitter, so that caveat is totally acceptable right now. Would love to make it support that though, for sure. Some stuff in there is still a bit crufty and outdated, but I wanted to start somewhere :)

nlf replied:
I totally agree, for now an EventEmitter is a simple way to get the functionality working and out there. A pipable stream can come later.

pgte commented on index.js:
```diff
@@ -152,8 +164,19 @@ RiakPBC.prototype.setBucket = function (params, callback) {
+    } 
+
+    if (streaming) {
+      var emitter = new EventEmitter();
```
Should be instantiating a readable stream:

```
var stream = new Readable({objectMode: true});
```

pgte commented on index.js:
```diff
@@ -168,8 +191,19 @@ RiakPBC.prototype.del = function (params, callback) {
+    } 
+
+    if (streaming) {
+      var emitter = new EventEmitter();
```
same here as above.

pgte commented on index.js:
```diff
@@ -82,9 +83,20 @@ function RiakPBC(options) {
+            var response = self.translator.decode(mc, packet.slice(1));
+
+            if (self.task.emitter && !response.done) {
+                self.task.emitter.emit('data', response);
```
should be something like:

``` javascript
self.task.stream.push(...
```

nlf commented:
This looks ready to me, merging and publishing version 0.0.10

roidrage commented:
Awesome, thank you!

I've already included this in my protobuffs branch for riak-js, which is now pretty much ready to be shipped! :)
